store: add initial symbol tables support

GiedriusS opened the pull request:
Add initial support for symbol tables; symbol tables are sent via hints that are used to deduplicate strings. The main benefit of doing this is to reduce network traffic && reduce number of allocations needed for all of the different strings.

Strings are referenced by their number. In total, there could be math.MaxUint64 strings. To avoid blocking in querier, symbol tables / references are automatically adjusted depending on which store it is in the stores slice. In other words, the whole math.MaxUint64 space is divided into equal number of strings for each StoreAPI. If that limit is breached then raw strings are sent instead. This is taken care of by the LookupTable builder. We ensure that old versions are compatible with this one by passing over the maximum number of allowed unique strings via StoreRequest. If that limit is zero then we disable the building of the lookup table.

This compression is beneficial in almost all cases. The worst case is when there are a lot of unique metrics with unique strings in each metric. However, I strongly believe that this will only happen 1% of the time due to the nature of monitoring. This is because a set of metrics will always have some identical dimensions i.e. same labels with only one or two changing.

I have also attempted an alternative implementation whereas a `Label` could be `oneof` compressed labels or the regular labels. However, the implementation wasn't quite as elegant, cumbersome. This is much cleaner.

For now, only streaming Sidecar + Thanos Store is implemented. We could add support for this in other components as a follow-up.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

fpetkovski commented:
I made a quick first pass, this looks pretty cool! I could not understand if the layered querier topology is covered as well. Is the table (indexing space) somehow recursively divided?

saswatamcode commented:
This looks really amazing! Would be interesting to see benchmarks too, especially with the lookup table bimap!

GiedriusS commented:
> I made a quick first pass, this looks pretty cool! I could not understand if the layered querier topology is covered as well. Is the table (indexing space) somehow recursively divided?

Each upper layer divides the uint64 space into equal parts for each StoreAPI. If any StoreAPI exceeds that limit then it errors out. Given that string references are unique, I think the function that I'm using should always remap all IDs properly.

I also thought about adding this behind a flag but is it realistic to have thousands and thousands of StoreAPIs that give millions of unique strings? :thinking: math.MaxUint64 is `18446744073709551615` :smile: I don't think it should even be reached. An even then maybe we could bump to uint128

saswatamcode commented:
> What benchmarks would you like to see for this? If this code looks good to my team mates then I'll try this out in prod and see the results 

I was thinking benchmarking the compression/decompression might be useful so that if this is changed in the future, we can catch regressions! 🙂 
But yes, it won't capture the network bandwidth benefits, so trying in prod sounds better for that.

GiedriusS commented:
Here's how network traffic dropped after deploying this:
![image](https://user-images.githubusercontent.com/2377233/203260816-83a4306c-0a60-407d-8169-56eb532d034e.png)

Dropped by 3x. This is one of the servers that is only evaluating rules via Thanos Ruler so the effect is visible the most.

fpetkovski commented:
Was there any reduction in query latency?

GiedriusS commented:
> Was there any reduction in query latency?

Doesn't look so, maybe just a little bit :( but with much smaller network usage, it's possible to stuff more recording/alerting rules into it

GiedriusS commented:
RAM usage also dropped around ~20% with no noticeable increase in CPU usage :thinking:  

![image](https://user-images.githubusercontent.com/2377233/203300286-94d75311-5f60-4e19-886e-eb5610a6b0a7.png)
![image](https://user-images.githubusercontent.com/2377233/203300333-1bd3f9aa-a16c-4ca3-9abf-b42e26f5234e.png)

bwplotka commented:
cc @bboreham - want to check it? You might have some thoughts. (:

matej-g commented on pkg/store/lookup_table.go:
```diff
@@ -0,0 +1,68 @@
+
+var maxElementsReached = errors.New("max elements reached in lookup table builder")
+
+func (b *lookupTableBuilder) putString(s string) (uint64, error) {
```
Nit: Maybe it would also then make sense to rename to something like `getOrStoreString`, since we're not really updating (putting) existing string.

matej-g commented on pkg/store/bucket.go:
```diff
@@ -1083,6 +1083,8 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		}
 	}
 
+	lookupTable := newLookupTableBuilder(req.MaximumStringSlots)
```
Nit: I'd keep it consistent and either call it `lookupTable` or `lookupTableBuilder` everywhere.

bboreham commented:
Thanks for this; I have considered doing something like this in the remote-write protocol.
A question that occurs to me is "when is a symbol table discarded from memory?", but perhaps I didn't understand.

Re some conversations we had at PromCon, you may like https://github.com/grafana/mimir-prometheus/pull/353.

matej-g reviewed (commented):
Nice job, the implementation looks good to me and your prod test results seem persuasive :eyes: , I have mostly small nits.

matej-g commented:
FWIW, idea somewhat related idea to others in this conversation - reusing strings in receiver https://github.com/thanos-io/thanos/pull/5899 - did not manage to finalize it yet though

cstyan commented:
Thanks @bboreham sending me this way.

@GiedriusS I have a similar implementation of a symbols table, based originally on the interner we already use in remote write. Let me know if you'd like to contribute yours upstream so that it could be reused in remote write.

GiedriusS commented on pkg/store/lookup_table.go:
```diff
@@ -0,0 +1,68 @@
+		return num, nil
+	}
+
+	s = strings.Clone(s)
```
Yep :+1:

GiedriusS commented:
Thanks for your review @matej-g, I have made adjustments according to your comments. Also, I have fixed one bug I noticed after deploying this in prod.

@bboreham the symbol table is built and discarded after a Series() call or, in other words, its lifetime/scope is the Select() call from the promql engine. I hope this answers your question. Perhaps we could improve this later on.

@cstyan by "upstream" do you mean Prometheus? Could you please post some links? Regardless, I'd prefer to merge this as-is and then we can work on further items.

fpetkovski commented on pkg/store/proxy_heap.go:
```diff
@@ -462,6 +463,49 @@ func newLazyRespSet(
 					seriesStats.Count(resp.GetSeries())
 				}
 
+				if resp.GetCompressedSeries() != nil {
```
I wonder if we can hide this behind some abstraction that knows whether to decompress the response or not. Maybe `GetSeries()` itself can do this if we pass in the hints?

fpetkovski replied:
Makes sense. I forgot that we need to wait until the end to get the table.

GiedriusS commented:
Been running this for months, haven't spotted any problems. PTAL.

fpetkovski commented on pkg/query/querier.go:
```diff
@@ -208,6 +211,47 @@ type seriesServer struct {
+		for _, cLabel := range cs.Labels {
+			var name, val string
+			for _, symTable := range s.symbolTables {
+				if foundName, ok := symTable[uint64(cLabel.NameRef)]; ok {
+					name = foundName
+				}
+
+				if foundValue, ok := symTable[uint64(cLabel.ValueRef)]; ok {
+					val = foundValue
+				}
```
Do we need to break the look if we found both name and value?

GiedriusS replied:
:+1: good point, will add

fpetkovski commented on pkg/store/storepb/rpc.proto:
```diff
@@ -122,6 +122,12 @@ message SeriesRequest {
+  // that the receiver can send. It is safe to ignore this by
+  // the receiver if it does not intend on sending a symbol table
+  // via hints later on.
+  uint64 maximum_string_slots = 14;
```
Why do we need to set a maximum limit here?

GiedriusS replied:
Because symbol references are uint64 and because a Querier might be connected to N different StoreAPIs, each sending references it means that it could use up to `math.Uint64Max/N` references. Without this limit, a Querier might not be able to proxy all of those references it receives to another querier.

fpetkovski replied:
Got it, makes sense 👍

GiedriusS commented:
Updated PR - merged `main` && made changes according to comments. Also, while running this in prod I've noticed that decompression takes a long time with queries that retrieve millions of series hence I've added concurrent decompression support that is tunable via a new flag. Also, now this functionality is disabled unless that flag is set to a value bigger than zero.

I can't imagine running Thanos in prod without this anymore because Queriers would get easily overwhelmed - we have thousands of recording rules and the labels completely dominate the % of bandwidth.

fpetkovski reviewed (approved):
LGTM!

Any recommendations on what value to set the number of workers to?

yeya24 commented:
How much does this feature help, any dashboards/metrics to share? @GiedriusS 👀

matej-g reviewed (approved):
⭐ ⭐ ⭐ 

One more nit regarding flag from me, but I'll leave it up to you, the implementation looks good 🙌

matej-g commented:
@GiedriusS anything still blocking this? Or you just need to resolve conflicts?

GiedriusS commented:
This will need to be reworked a bit after the latest changes because order became important now. It means that we need to somehow merge different symbol references from multiple StoreAPIs in such a way as to preserve order. I don't know how to do that, at the moment. The references will need to somehow be encoded in such a way as to preserve order.

yeya24 commented:
Do we still want to continue this work?

MitchellJThomas commented:
+1 for continuing the work

ringerc commented:
I have hesitations about adopting something that makes streaming large result sets even harder, personally. This is often desirable for intermediate results in distributed Query etc.

I'd also be worried about the impact this has on memory use. How and when is the symbol table pruned, and how is its memory capped + made visible for monitoring?

Nothing wrong with it being optional.

Have you considered a streaming friendly dictionary encoder with a LRU to cap memory use? It would only help with repeated labels within one response, not between responses, but should have significant benefits.
